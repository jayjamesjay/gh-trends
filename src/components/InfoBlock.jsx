import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { languages as colors, RepoInfo } from './Data';
import { TextBlock } from '../styles/Main';
import Article from '../styles/Article';
import Span from '../styles/Span';
import { H2 } from '../styles/Headers';
import { Img, ImgInline } from '../styles/Img';
import { P, PAlt, PClean } from '../styles/Paragraph';
import { ButtonAdd } from '../styles/Button';
import { LinkA } from '../styles/Link';

// Block of information about repository with an indicator if it was saved
export default class InfoBlock extends Component {
  save = () => {
    const { info, save } = this.props;
    save(info);
  };

  render() {
    const {
      props: { info, saved },
      save
    } = this;
    const { language, stargazersCount } = info;
    const license = info.license ? `License: ${info.license}` : '';
    const forks = `Forks: ${info.forks}`;
    let color;

    if (Object.prototype.hasOwnProperty.call(colors, language)) {
      color = colors[language];
    } else {
      color = null;
    }

    return (
      <Article bg={color}>
        <H2>
          <RepoLink url={info.url} nameWithOwner={info.nameWithOwner} />
        </H2>
        <P>{info.description}</P>
        <PClean>{forks}</PClean>
        <PClean>{license}</PClean>
        <SaveRepo save={save} saved={saved} />
        <TextBlock>
          <PAlt>
            {stargazersCount}
            <ImgInline src="./assets/img/stars.svg" alt="Stars" />
          </PAlt>
          <PAlt>{language}</PAlt>
        </TextBlock>
      </Article>
    );
  }
}

InfoBlock.propTypes = {
  saved: PropTypes.bool.isRequired,
  save: PropTypes.func.isRequired,
  info: PropTypes.instanceOf(RepoInfo).isRequired
};

// Llink to repository
export function RepoLink({ url, nameWithOwner }) {
  const [author, name] = nameWithOwner.split('/');

  return (
    <LinkA href={url}>
      <Span>{author}</Span>
      {' /'}
      {name}
    </LinkA>
  );
}

RepoLink.propTypes = {
  url: PropTypes.string.isRequired,
  nameWithOwner: PropTypes.string.isRequired
};

// Button to add/remove repository from saved
export function SaveRepo({ save, saved }) {
  return (
    <ButtonAdd onClick={save}>
      {saved ? (
        <Img src="./assets/img/delete.svg" alt="Remove from saved" title="Remove from saved" />
      ) : (
        <Img src="./assets/img/add-saved.svg" alt="Add to saved" title="Add to saved" />
      )}
    </ButtonAdd>
  );
}

SaveRepo.propTypes = {
  saved: PropTypes.bool.isRequired,
  save: PropTypes.func.isRequired
};
