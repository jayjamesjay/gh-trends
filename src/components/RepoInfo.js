/**
 * @module RepoInfo
 */

/**
 * Information about repository
 * @class
 */
class RepoInfo {
  /**
   * Creates a RepoInfo
   *
   * @param {string} nameWithOwner - name and author of repository
   * @param {string} url - url of the repository
   * @param {string} description - description
   * @param {number} stargazersCount - amount of stars
   * @param {string} language - programming language
   * @param {number} forks - amount of forks
   * @param {string} license - repository license
   */
  constructor(nameWithOwner, url, description, stargazersCount, language, forks, license) {
    this.nameWithOwner = nameWithOwner;
    this.url = url;
    this.description = description;
    this.stargazersCount = stargazersCount;
    this.language = language;
    this.forks = forks;
    this.license = license;
  }

  /**
   * Creates RepoInfo from Github API data
   *
   * @param {object} item - repository data from Github API
   * @returns {RepoInfo}
   */
  static fromGithubRes(item) {
    const { description, language, forks } = item;
    const license = item.license ? item.license.spdx_id : null;

    return new RepoInfo(
      item.full_name,
      item.html_url,
      description,
      item.stargazers_count,
      language,
      forks,
      license
    );
  }
}

export default RepoInfo;
