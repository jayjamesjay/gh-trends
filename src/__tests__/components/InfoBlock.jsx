import React from 'react';
import 'jest-styled-components';
import { render, screen } from '@testing-library/react';
import InfoBlock, { RepoLink, SaveRepo } from '../../components/InfoBlock';
import RepoInfo from '../../utils/RepoInfo';

describe('<InfoBlock />', () => {
  const info = new RepoInfo(
    'jayjamesjay/gh-trends',
    '',
    'Loading content for this website...',
    123,
    'JavaScript',
    321,
    'MIT',
  );
  const saved = false;
  const save = (currInfo) => currInfo;

  it(`renders with default style`, () => {
    const { asFragment } = render(<InfoBlock info={info} save={save} saved={saved} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with license', () => {
    render(<InfoBlock info={info} save={save} saved={saved} />);
    expect(screen.queryByText(`License: ${info.license}`)).not.toBeNull();
  });

  it('renders without license', () => {
    info.license = null;

    render(<InfoBlock info={info} save={save} saved={saved} />);
    expect(screen.queryByText(`License: ${info.license}`)).toBeNull();
  });
});

describe('<RepoLink />', () => {
  const author = 'jayjamesjay';
  const name = 'gh-trends';
  const title = `${author}/${name}`;
  const url = 'https://github.com/jayjamesjay/gh-trends';

  it(`renders with default style`, () => {
    const { asFragment } = render(<RepoLink nameWithOwner={title} url={url} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays owner', () => {
    render(<RepoLink nameWithOwner={title} url={url} />);
    expect(screen.queryByText(author).textContent).not.toBeNull();
  });

  it('displays name', () => {
    render(<RepoLink nameWithOwner={title} url={url} />);
    expect(screen.queryByText(`/${name}`).textContent).not.toBeNull();
  });
});

describe('<SaveRepo />', () => {
  const func = () => {};
  const btn = (saved) => render(<SaveRepo save={func} saved={saved} />);

  it(`renders with default style`, () => {
    const { asFragment } = render(<SaveRepo save={func} saved />);
    expect(asFragment()).toMatchSnapshot();
  });

  it(`displays add icon if repo hasn't been saved`, () => {
    const saved = false;
    btn(saved);

    expect(screen.getByRole('img').getAttribute('alt')).toContain('Add');
  });

  it(`displays remove icon if repo has been saved`, () => {
    const saved = true;
    btn(saved);

    expect(screen.getByRole('img').getAttribute('alt')).toContain('Remove');
  });
});
