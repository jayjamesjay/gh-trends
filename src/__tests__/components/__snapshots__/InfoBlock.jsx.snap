// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoBlock /> renders with default style 1`] = `
.c2 {
  display: block;
  font-size: 1.2rem;
  font-family: 'Noto Sans HK',sans-serif;
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: row wrap;
  -ms-flex-flow: row wrap;
  flex-flow: row wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  margin: auto 0 0 0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 2rem 1rem;
  min-height: 13rem;
  padding: 1.2rem 1.2rem 0.8rem 1.2rem;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  background: linear-gradient( #fff 0%,#fff 0.3rem,#f1e05a 0.3rem,#f1e05a 0.6rem,#fff 0.6rem );
  box-shadow: 0px 3px 5px 1px rgba(217,217,217,0.82);
  border-radius: 0.5rem;
  text-align: left;
  word-break: break-word;
}

.c3 {
  font-weight: 400;
  font-size: 1.3rem;
  color: #767676;
}

.c1 {
  margin: 0.2rem 0;
  font-size: 1.7rem;
}

.c8 {
  width: 100%;
  height: auto;
  -webkit-filter: invert(0%);
  filter: invert(0%);
}

.c11 {
  display: inline-block;
  width: auto;
  height: 2rem;
  -webkit-filter: invert(0%);
  filter: invert(0%);
}

.c9 {
  margin: 0 0.5rem;
  padding: 0;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0 0.5rem;
  padding: 0;
}

.c6 {
  cursor: pointer;
  height: auto;
  background: transparent;
  border: 0;
}

.c7 {
  margin: 0.5rem 0 0 0;
}

@media screen and (min-width:40rem) {
  .c0 {
    min-width: 16.5rem;
    margin: 1.5rem 1rem;
    -webkit-flex: 1 1 27%;
    -ms-flex: 1 1 27%;
    flex: 1 1 27%;
  }
}

<article
  className="c0"
>
  <h2
    className="c1"
  >
    <a
      className="c2"
      href=""
    >
      <span
        className="c3"
      >
        jayjamesjay
      </span>
       /
      gh-trends
    </a>
  </h2>
  <p>
    Loading content for this website...
  </p>
  <div
    className="c4 c5"
  >
    <button
      className="c6 c7"
      onClick={[Function]}
    >
      <img
        alt="Add to saved"
        className="c8"
        src="test-file-stub"
        title="Add to saved"
      />
    </button>
    <p
      className="c9"
    >
      Forks: 321
    </p>
    <p
      className="c9"
    >
      License: MIT
    </p>
    <p
      className="c10"
    >
      123
      <img
        alt="Stars"
        className="c11"
        src="test-file-stub"
      />
    </p>
    <p
      className="c9"
    >
      Language: JavaScript
    </p>
  </div>
</article>
`;

exports[`<RepoLink /> renders with default style 1`] = `
.c0 {
  display: block;
  font-size: 1.2rem;
  font-family: 'Noto Sans HK',sans-serif;
  color: #000;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c1 {
  font-weight: 400;
  font-size: 1.3rem;
  color: #767676;
}

<a
  className="c0"
  href="https://github.com/jayjamesjay/gh-trends"
>
  <span
    className="c1"
  >
    jayjamesjay
  </span>
   /
  gh-trends
</a>
`;

exports[`<SaveRepo /> renders with default style 1`] = `
.c2 {
  width: 100%;
  height: auto;
  -webkit-filter: invert(0%);
  filter: invert(0%);
}

.c0 {
  cursor: pointer;
  height: auto;
  background: transparent;
  border: 0;
}

.c1 {
  margin: 0.5rem 0 0 0;
}

<button
  className="c0 c1"
  onClick={[Function]}
>
  <img
    alt="Remove from saved"
    className="c2"
    src="test-file-stub"
    title="Remove from saved"
  />
</button>
`;
